internal/pkg/agent/application/upgrade/details/details.go

```diff
@@ -0,0 +1,95 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package details
+
+import "sync"
+
+// Observer is a function that will be called with upgrade details
+type Observer func(details *Details)
+
+// Details consists of details regarding an ongoing upgrade.
+type Details struct {
+	TargetVersion string          `yaml:"target_version" json:"target_version"`
+	State         State           `yaml:"state" json:"state"`
+	ActionID      string          `yaml:"action_id" json:"action_id"`
+	Metadata      DetailsMetadata `yaml:"metadata" json:"metadata"`
+
+	observers []Observer
+	mu        sync.RWMutex
+}
+
+// DetailsMetadata consists of metadata relating to a specific upgrade state
+type DetailsMetadata struct {
+	ScheduledAt     string  `yaml:"scheduled_at" json:"scheduled_at"`
+	DownloadPercent float64 `yaml:"download_percent" json:"download_percent"`
+	FailedState     State   `yaml:"failed_state" json:"failed_state"`
+	ErrorMsg        string  `yaml:"error_msg" json:"error_msg"`
+}
+
+func NewDetails(targetVersion string, initialState State, actionID string, metadata DetailsMetadata) *Details {
+	return &Details{
+		TargetVersion: targetVersion,
+		State:         initialState,
+		ActionID:      actionID,
+		Metadata:      metadata,
+		observers:     []Observer{},
+	}
+}
+
+// SetState is a convenience method to set the state of the upgrade and
+// notify all observers.
+func (d *Details) SetState(s State) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	d.State = s
+	d.notifyObservers()
+}
+
+// Fail is a convenience method to set the state of the upgrade
+// to StateFailed, set metadata associated with the failure, and
+// notify all observers.
+func (d *Details) Fail(err error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	d.Metadata.FailedState = d.State
+	d.Metadata.ErrorMsg = err.Error()
+	d.State = StateFailed
+	d.notifyObservers()
+}
+
+// RegisterObserver allows an interested consumer of Details to register
+// themselves as an Observer. The registered observer is immediately notified
+// of the current upgrade details.
+func (d *Details) RegisterObserver(observer Observer) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	d.observers = append(d.observers, observer)
+	d.notifyObserver(observer)
+}
+
+// NotifyObservers sends the current upgrade details to all registered
+// observers.
+func (d *Details) NotifyObservers() {
+	d.mu.RLock()
+	defer d.mu.RUnlock()
```

Comment:
There could be multiple observers registered.  In such a case they would be called serially, in the loop inside the `notifyObservers` method.

However, even if there was only one single observer registered, the `RLock` call is needed to ensure that the internal state of `d`, e.g. `d.State` or `d.Metadata`, haven't changed between the caller calling `NotifyObservers()` and eventually the observer function being called inside the `notifyObserver` method.  Also, I used an `RLock` instead of a `Lock` because I'm okay with multiple reads of `d`'s internal state while `NotifyObservers()` is executing; I just don't want a write to sneak in there as explained earlier.

Reply:
The goal I'm trying to achieve here is that for a given `UpgradeDetails` object, `d`, I want all it's observers to see the same value of `d` when `d.NotifyObservers()` is called.  Because an `RLock()` would prevent concurrent _writes_ to the fields of `d`, I think my goal is achieved.  

You're right that a regular `Lock()` would also achieve that goal but it would be a stronger lock than I need. I don't necessarily care if there are two concurrent invocations of `d.NotifyObservers()` for the same `d` object; in fact, I _would_ want both to be executed without the first one blocking the second one, which would happen if we used a `Lock()` here.  

> I think I am missing something here as I don't see other functions holding an RLock for reading state. If those reads happen without any lock at all, again there's no difference between RLock and a normal mutex.

Are you referring to `notifyObservers()` and `notifyObserver()`?  I did not use any mutexes in there because otherwise we end up with a deadlock when `d.SetState()` is called.  These are private methods and the public methods that call them acquire/release the mutex.

Reply:
> Multiple RLock() can be acquired simultaneously so in the current code the notification would run concurrently. Is this the desired result ? 

Yes, as explained in the previous comment.

> If that's the case, did we write down somewhere that the Observers' code MUST be thread safe ?

No, but you bring up an excellent point. Observer's should receive a (deep) copy of the details object, not a pointer to it.  The idea is that the observer receives a "snapshot-in-time" of the upgrade details at the time of observation.  I have fixed this now in 5bbb507ee.

Reply:
It's not just in the manipulation of the UpgradeDetails, it's a warning for the implementation of the observers to protect their own state against concurrent calls...

Reply:
> It's not just in the manipulation of the UpgradeDetails, it's a warning for the implementation of the observers to protect their own state against concurrent calls...

Now I think I might be missing something 😄.  Not a rhetorical question: why should we add such a warning?  Isn't it generally true that any object that wants to support concurrent access may want to protect its own state?